dirty_cat/gamma_poisson_factorization.py

```diff
@@ -0,0 +1,600 @@
+
+    rescale_W : boolean, default=True
+        If true, the weight matrix W is rescaled at each iteration
+        to have an l1 norm equal to 1 for each row.
+    
+    max_iter_e_step : int, default=20
+        Maximum number of iterations to adjust the activations h at each step.
+
+
+    Attributes
+    ----------
+
+    References
+    ----------
+    For a detailed description of the method, see
+    `Encoding high-cardinality string categorical variables
+    <https://hal.inria.fr/hal-02171256v4>`_ by Cerda, Varoquaux (2019).
+    """
+
+    def __init__(self, n_topics=10, batch_size=512, gamma_shape_prior=1.1,
+                 gamma_scale_prior=1.0, rho=.95, rescale_rho=False,
+                 hashing=False, hashing_n_features=2**12, init='k-means++',
+                 tol=1e-4, min_iter=2, max_iter=5, ngram_range=(2, 4),
+                 analyzer='char', add_words=False, random_state=None,
+                 rescale_W=True, max_iter_e_step=20):
+
+        self.ngram_range = ngram_range
+        self.n_topics = n_topics
+        self.gamma_shape_prior = gamma_shape_prior  # 'a' parameter
+        self.gamma_scale_prior = gamma_scale_prior  # 'b' parameter
+        self.rho = rho
+        self.rescale_rho = rescale_rho
+        self.batch_size = batch_size
+        self.tol = tol
+        self.hashing = hashing
+        self.hashing_n_features = hashing_n_features
+        self.max_iter = max_iter
+        self.min_iter = min_iter
+        self.init = init
+        self.analyzer = analyzer
+        self.add_words = add_words
+        self.random_state = check_random_state(random_state)
+        self.rescale_W = rescale_W
+        self.max_iter_e_step = max_iter_e_step
+
+        # Init n-grams counts vectorizer
+        if self.hashing:
+            self.ngrams_count = HashingVectorizer(
+                 analyzer=self.analyzer, ngram_range=self.ngram_range,
+                 n_features=self.hashing_n_features,
+                 norm=None, alternate_sign=False)
+            if self.add_words: # Init a word counts vectorizer if needed
+                self.word_count = HashingVectorizer(
+                     analyzer='word',
+                     n_features=self.hashing_n_features,
+                     norm=None, alternate_sign=False)
+        else:
+            self.ngrams_count = CountVectorizer(
+                 analyzer=self.analyzer, ngram_range=self.ngram_range)
+            if self.add_words:
+                self.word_count = CountVectorizer()
+
+    def _update_H_dict(self, X, H):
+        """
+        For each category x in X, update the dictionary self.H_dict with
+        the corresponding bag-of-n-grams representation h.
+
+        Parameters
+        ----------
+        X : array-like, shape (n_samples, )
+            The string data to fit the model on.
+        H : array-like, shape (n_samples, n_vocab)
+            The corresponding bag-of-n-grams representations.
+        """
+        for x, h in zip(X, H):
```

Comment:
Using a list comprehension will be faster.

Reply:
What do you have in mind ? I cannot use assignments like `self.H_dict[x] = h` in a list comprehension, so something like `[self.H_dict[x] = h for x, h in zip(X, H)]` doesn't work.